tools/tpm2_sign.c

```diff
@@ -88,83 +88,93 @@ static tool_rc sign_and_save(ESYS_CONTEXT *ectx) {
 
 static tool_rc init(ESYS_CONTEXT *ectx) {
 
-    bool option_fail = false;
+    /*
+     * Set signature scheme for key type, or validate chosen scheme is
+     * allowed for key type.
+     */
+    tool_rc rc = tpm2_alg_util_get_signature_scheme(ectx,
+            ctx.signing_key.object.tr_handle, ctx.halg, ctx.sig_scheme,
+            &ctx.in_scheme);
+    if (rc != tool_rc_success) {
+        LOG_ERR("Invalid signature scheme for key type!");
+        return rc;
+    }
 
     if (ctx.cp_hash_path && ctx.output_path) {
-        LOG_ERR("Cannout output signature when calculating cpHash");
+        LOG_ERR("Cannot output signature when calculating cpHash");
         return tool_rc_option_error;
     }
 
     if (!ctx.signing_key.ctx_path) {
         LOG_ERR("Expected option c");
-        option_fail = true;
+        return tool_rc_option_error;
     }
 
     if (!ctx.flags.o && !ctx.cp_hash_path) {
         LOG_ERR("Expected option o");
-        option_fail = true;
-    }
-
-    if (option_fail) {
         return tool_rc_option_error;
     }
 
-    if (ctx.flags.d && ctx.flags.t) {
-        LOG_WARN("When using a pre-computed digest the validation ticket"
-                " is ignored.");
-    }
-
-    if (ctx.flags.d || !ctx.flags.t) {
-        ctx.validation.tag = TPM2_ST_HASHCHECK;
-        ctx.validation.hierarchy = TPM2_RH_NULL;
-        memset(&ctx.validation.digest, 0, sizeof(ctx.validation.digest));
+    if (!ctx.flags.d && ctx.flags.t) {
+        LOG_WARN("Invalid validation ticket since no digest specified.");
     }
 
     /*
-     * Set signature scheme for key type, or validate chosen scheme is allowed for key type.
+     * Applicable when input data is not a digest, rather the message to sign.
+     * A digest is calculated first in this case.
      */
-    tool_rc rc = tpm2_alg_util_get_signature_scheme(ectx,
-            ctx.signing_key.object.tr_handle, ctx.halg, ctx.sig_scheme,
-            &ctx.in_scheme);
-    if (rc != tool_rc_success) {
-        LOG_ERR("bad signature scheme for key type!");
-        return rc;
-    }
-
-    /* Process the msg file if needed */
     if (!ctx.flags.d) {
         FILE *input = ctx.input_file ? fopen(ctx.input_file, "rb") : stdin;
         if (!input) {
             LOG_ERR("Could not open file \"%s\"", ctx.input_file);
             return tool_rc_general_error;
         }
 
-        rc = tpm2_hash_file(ectx, ctx.halg, TPM2_RH_NULL, input, &ctx.digest,
-                NULL);
+        TPMT_TK_HASHCHECK *temp_validation_ticket;
+        rc = tpm2_hash_file(ectx, ctx.halg, TPM2_RH_OWNER, input, &ctx.digest,
+                &temp_validation_ticket);
```

Comment:
Does this always work, even if the object is in a different hierarchy? It either has to always work, or we leave this advanced feature for discrete tpm2_hash calls and clearly spell this out in the manpage.

Reply:
It does always work. Although I really don't like duplicating the hashing code in the signing tool. Perhaps it still needs to be around for compatibility reasons. What are your thoughts on dropping the hashing mechanisms from signing tool and only sign digests with or without validation tickets?

Reply:
It needs to be around for compatibility reasons, if we make the default behavior a digest and not a message, it would break all existing callers (5.0 release). I would just leave it, set it as owner hierachy **and document it in the manpage**. So folks know if they want more features they need to call tpm2_hash and then tpm2_sign for full control.